### PR TITLE
[processing] 2 more plots added to the collection

### PR DESCRIPTION
**Pull request by ghtmtt**
This PR re-uses the code of VectorLayerScatterPlot to make a 3D plot  and adds a simple BoxPlot

### CODE REVIEW

**Comment by m-kuhn** [comment]
Do you think you could add unit tests for these?

Something like https://github.com/qgis/QGIS/blob/master/python/plugins/processing/tests/testdata/qgis_algorithm_tests.yaml#L1814-L1824

**Comment by ghtmtt** [comment]
mmm.. really don't not how to deal with the html output file and run an unit test..  the html file is super huge and with a lot of default values and strings.. (e.g. `!function(t){if("object"==typeof exports&&"undefined"!=typeof module)module.exports=t()` and it continues for thousands of lines).. do you have an idea?

**Comment by ghtmtt** [comment]
@alexbruy I noticed something wrong with the x variable in the BarPlot (related to numpy, the resulting x axis does not show the correct range).
I also noticed that strings were not allowed in the x variable (and discovered that the method of Processing https://github.com/qgis/QGIS/blob/master/python/plugins/processing/tools/vector.py#L190 automatically converts fields into floats).
Could you have a look at the super small fixes made?

Thanks!

**Comment by m-kuhn** [comment]
You could look for some strings which are visible in the end result (like axis labels etc) and regex for them. Just some general safety check if there is output generated and if it contains some of the desired result. This will at least control that a) the algorithm runs fine (no exceptions thrown etc), b) there is an output produced and c) that the output contains "something probably useful".

**Comment by ghtmtt** [comment]
Totally agree with a kind of general check that assures the output. I will try to do something.. Let you now

**Comment by ghtmtt** [comment]
@m-kuhn  I was not able to write a test, sorry.. can we merge it anyway? I'll look to write some test..
or better, @alexbruy can we use your new https://github.com/qgis/QGIS/pull/4268 test suite for the plots?

**Comment by ghtmtt** [comment]
@m-kuhn what do you think about this PR? even without tests, could we merge it (after having resolved the conflicts)?

File: python/plugins/processing/algs/qgis/BoxPlot.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+        self.addParameter(ParameterTableField(self.NAME_FIELD,
+                                              self.tr('Category name field'),
+                                              self.INPUT,
+                                              ParameterTableField.DATA_TYPE_ANY))
+        self.addParameter(ParameterTableField(self.VALUE_FIELD,
+                                              self.tr('Value field'),
+                                              self.INPUT,
+                                              ParameterTableField.DATA_TYPE_NUMBER))
+        self.addParameter(ParameterBoolean(self.MSD,
+                                           self.tr('Show also standard deviation'), False))
```
Comment (m-kuhn) [inline python/plugins/processing/algs/qgis/BoxPlot.py]: Should that be a `ParameterSelection` with options `True`, `False`, `'sd'`?

File: python/plugins/processing/algs/qgis/BoxPlot.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+                    x=x_var,
+                    y=values[valuefieldname],
+                    boxmean=True)]
+        else:
+            data = [go.Box(
+                    x=x_var,
+                    y=values[valuefieldname],
+                    boxmean='sd')]
+
+        plt.offline.plot(data, filename=output, auto_open=False)
```
Comment (m-kuhn) [inline python/plugins/processing/algs/qgis/BoxPlot.py]: Would it be interesting to add `auto_open` as a parameter?

**Comment by m-kuhn** [comment]
Sure, this is mergable without tests.
Just ping me when you have it rebased (and handled the comments if you think they make sense).

File: python/plugins/processing/algs/qgis/BoxPlot.py (hunk 1 above)
Comment (ghtmtt) [inline python/plugins/processing/algs/qgis/BoxPlot.py]: yes I should be.. I tried to follow some *templates* but it turn out that when looping on the `ParameterSelection` within the script, then I get the indexes of the parameters and not the parameters itself (needed by `plotly`). So in the meanwhile to look for something smarter I was thinking to let the user choose between `yes` or `no`

File: python/plugins/processing/algs/qgis/BoxPlot.py (hunk 2 above)
Comment (ghtmtt) [inline python/plugins/processing/algs/qgis/BoxPlot.py]: that is also the default parameter.. but I saw that the first plots added bu @alexbruy have the `auto_open` set to False. If also Alex agrees, I can add a checkbox (maybe default checked?!) with the `auto_open` option

File: python/plugins/processing/algs/qgis/BoxPlot.py (hunk 1 above)
Comment (m-kuhn) [inline python/plugins/processing/algs/qgis/BoxPlot.py]: Just do

```
msdIndex = self.getParameterValue(self.MSD)
msd = True
if msdIndex == 1:
    msd = 'sd'
elif msdIndex == 2:
   msd = False
```

Or if you want to do it clean and fancy, define an enum somewhere on top of your file

```
MSD_SD, MSD_ON, MSD_OFF = range(3)
```

and then use

```
if msdIndex == MSD_ON:
```

etc.

------------

Sidenote, I would skip the word `also` - and simply say "Show standard deviation"

**Comment by ghtmtt** [comment]
ok, done ;) thanks for the hints of the `combobox` in Processing.. 
I rebased and fixed the damns conflicts... hopefully travis won't complain

Thanks again @m-kuhn

**Comment by m-kuhn** [comment]
Whoops, look like something has gone a bit weird here...

**Comment by ghtmtt** [comment]
@m-kuhn as always :( ... could you please trash this PR? I made a new one from scratch #4359 cleaner and simpler..

**Comment by m-kuhn** [comment]
If you paste here what you tried, I can try to tell you where things went wrong...

**Comment by ghtmtt** [comment]
I thought it could be simple...
* git checkout master
* git pull upstream master
* git checkout proc_plotly
* git rebase master

fixing some conflicts and then just as always (git add, git commit.....).

But don't waste your time...

**Comment by m-kuhn** [comment]
I guess your local master is somehow not in sync with the upstream master.

To bypass the problem (next time) use this instead:

```
git checkout proc_plotly
git fetch upstream
git rebase upstream/master
```

Then repeat

```
git mergetool
git rebase --continue
# Or sometimes ...
git rebase --skip
```

**Comment by ghtmtt** [comment]
thanks! Next time I'll try this ;)

**Pull request closed without merging**